clmock/clmock.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package clmock
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type CLMock struct {
+	ctx               context.Context
+	cancel            context.CancelFunc
+	stack             *node.Node
+	eth               *eth.Ethereum
+	blockPeriod       time.Duration
+	withdrawals       []*types.Withdrawal
+	feeRecipient      common.Address
+	feeRecipientMutex sync.Mutex
+	nextWithdrawalIdx uint64
+}
+
+func NewCLMock(stack *node.Node, eth *eth.Ethereum) *CLMock {
+	chainConfig := eth.APIBackend.ChainConfig()
+
+	if chainConfig.Dev == nil {
+		log.Crit("incompatible pre-existing chain configuration")
+	}
+
+	return &CLMock{
+		stack:        stack,
+		eth:          eth,
+		blockPeriod:  time.Duration(chainConfig.Dev.Period) * time.Second,
+		withdrawals:  []*types.Withdrawal{},
+		feeRecipient: common.Address{},
+	}
+}
+
+// Start invokes the clmock life-cycle function in a goroutine
+func (c *CLMock) Start() error {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.clmockLoop()
+	return nil
+}
+
+// Stop halts the clmock service
+func (c *CLMock) Stop() error {
+	c.cancel()
+	return nil
+}
+
+func (c *CLMock) addWithdrawal(w types.Withdrawal) error {
+	if w.Index < c.nextWithdrawalIdx {
+		return fmt.Errorf("withdrawal has index (%d) less than or equal to latest received withdrawal index (%d)", w.Index, c.nextWithdrawalIdx-1)
+	}
+	c.nextWithdrawalIdx = w.Index + 1
+	c.withdrawals = append(c.withdrawals, &w)
```

Comment:
i think that these two lines need to be done with the mutex? otherwise might be a race between the api and building blocks?

Reply:
I added a mutex lock for the entire function so that there isn't a race when reading the value of `c.nextWithdrawalIdx` above.